lib/posix/pthread_key.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <semaphore.h>
+#include <errno.h>
+#include <posix/pthread.h>
+
+static sem_t sem;
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *newKey;
+
+	*key = NULL;
+
+	sem_init(&sem, 0, 1);
+
+	newKey = (pthread_key_obj *)k_malloc(sizeof(pthread_key_obj));
+
+	if (newKey == NULL) {
+		errno = ENOMEM;
+		return -1;
+	}
+
+	sys_dlist_init(&(newKey->KeyDataL));
+
+	newKey->destructor = destructor;
+	*key = (void *)newKey;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *keyObj = (pthread_key_obj *)key;
+	KeyData *keyData;
+
+	sem_wait(&sem);
+
+	/* Delete thread-specific elements associated with the key */
+	while (!sys_dlist_is_empty(&(keyObj->KeyDataL))) {
```

Comment:
Also, while it's fine to pass `NULL` to `k_free()`, but since `sys_dlist_get()` will call `sys_dlist_is_empty()` before attempting to dereference `list`, you might want to use `SYS_DLIST_FOR_EACH_NODE_SAFE`; for instance, take this example from `mem_domain.c`:

```C
        SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
                struct k_thread *thread =
                        CONTAINER_OF(node, struct k_thread, mem_domain_info);

                sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
                thread->mem_domain_info.mem_domain = NULL;
        }
```

Reply:
Oh alright, I will use this macro instead to loop over the list elements.

So then my loop would be

    SYS_DLIST_FOR_EACH_NODE_SAFE(&(key_obj->key_data_l), node_l, node_next_l) {
        key_data = (pthread_key_data *)sys_dlist_get(&(key_obj->key_data_l));
        k_free((void *)key_data);
    }
 
I would still call `sys_dlist_get()` though, to get the node, so I can free its memory. Is that okay? Or should I directly call `sys_dlist_remove()`?